lib/Sema/DerivedConformanceEquatableHashable.cpp

```diff
@@ -185,33 +389,217 @@ static void deriveBodyEquatable_enum_eq(AbstractFunctionDecl *eqDecl) {
+      TuplePatternElt(lhsElemPat), TuplePatternElt(rhsElemPat) },
+                                                 SourceLoc());
+    caseTuplePattern->setImplicit();
+
+    auto labelItem = CaseLabelItem(/*IsDefault*/ false, caseTuplePattern,
+                                   SourceLoc(), nullptr);
+
+    // Generate a guard statement for each associated value in the payload,
+    // breaking out early if any pair is unequal. (This is done to avoid
+    // constructing long lists of autoclosure-wrapped conditions connected by
+    // &&, which the type checker has more difficulty processing.)
+    SmallVector<ASTNode, 6> statementsInCase;
+    for (size_t varIdx = 0; varIdx < lhsPayloadVars.size(); varIdx++) {
+      auto lhsVar = lhsPayloadVars[varIdx];
+      auto lhsExpr = new (C) DeclRefExpr(ConcreteDeclRef(lhsVar), DeclNameLoc(),
+                                         /*implicit*/true);
+      auto rhsVar = rhsPayloadVars[varIdx];
+      auto rhsExpr = new (C) DeclRefExpr(ConcreteDeclRef(rhsVar), DeclNameLoc(),
+                                         /*Implicit*/true);
+      auto guardStmt = returnIfNotEqualGuard(C, lhsExpr, rhsExpr);
+      statementsInCase.emplace_back(guardStmt);
+    }
+
+    // If none of the guard statements caused an early exit, then all the pairs
+    // were true.
+    // return true
+    auto trueExpr = new (C) BooleanLiteralExpr(true, SourceLoc(),
+                                               /*Implicit*/true);
+    auto returnStmt = new (C) ReturnStmt(SourceLoc(), trueExpr);
+    statementsInCase.push_back(returnStmt);
+
+    auto body = BraceStmt::create(C, SourceLoc(), statementsInCase,
+                                  SourceLoc());
+    cases.push_back(CaseStmt::create(C, SourceLoc(), labelItem, hasBoundDecls,
+                                     SourceLoc(), body));
+  }
+
+  // default: result = false
+  //
+  // We only generate this if the enum has more than one case. If it has exactly
+  // one case, then that single case statement is already exhaustive.
+  if (elementCount > 1) {
+    auto defaultPattern = new (C) AnyPattern(SourceLoc());
+    defaultPattern->setImplicit();
+    auto defaultItem = CaseLabelItem(/*IsDefault*/ true, defaultPattern,
+                                     SourceLoc(), nullptr);
+    auto falseExpr = new (C) BooleanLiteralExpr(false, SourceLoc(),
+                                                /*implicit*/ true);
+    auto returnStmt = new (C) ReturnStmt(SourceLoc(), falseExpr);
+    auto body = BraceStmt::create(C, SourceLoc(), ASTNode(returnStmt),
+                                  SourceLoc());
+    cases.push_back(CaseStmt::create(C, SourceLoc(), defaultItem,
+                                     /*HasBoundDecls*/ false,
+                                     SourceLoc(), body));
+  }
+
+  // switch (a, b) { <case statements> }
+  auto aRef = new (C) DeclRefExpr(ConcreteDeclRef(aParam), DeclNameLoc(),
```

Comment:
The ConcreteDeclRef ctor is implicit so you don't have to call it explicitly (you rely on the implicit behavior below)

Reply:
Removed throughout.